### PR TITLE
Support bulk data export

### PR DESCRIPTION
**Pull request by demusdev**
## Description

Closes: #353 

## Checklist before requesting a review

- [x] Unit and integration tests added
  
- [ ] Compatibility:
    
  - [ ] Network APIs: ✅ - NO EFFECT
    
  - [ ] Workspace layout and metadata: ✅  - NO EFFECT
    
  - [ ] Configuration: ✅  - NO EFFECT
    
  - [ ] Container images: ✅  - NO EFFECT
- [ ] Observability:
    
  - [ ] Tracing / [metrics](https://github.com/kamu-data/kamu-standards/blob/master/metrics_design.md): ✅  - NO NEED
    
  - [ ] Alerts: ✅  - NO NEED
- [ ] Documentation:
    
  - [x] [Changelog](./CHANGELOG.md): ✅
    
  - [x] [Public documentation](https://github.com/kamu-data/kamu-docs/): ✅
  
- [ ] Downstream effects:
    
  - [ ] [kamu-node](https://github.com/kamu-data/kamu-node): ✅  - NO EFFECT
    
  - [ ] [kamu-web-ui](https://github.com/kamu-data/kamu-web-ui): ✅  - NO EFFECT
    
  - [ ] [release-queue](https://github.com/orgs/kamu-data/projects/4/views/1)

### CODE REVIEW

File: src/app/cli/src/cli.rs
Comment (sergiimk) [inline src/app/cli/src/cli.rs:381]: Better to use a more specific type

File: src/app/cli/src/cli.rs
Comment (sergiimk) [inline src/app/cli/src/cli.rs:390]: Unclear if it's in bytes or records

File: src/app/cli/src/commands/export_command.rs
Comment (sergiimk) [inline src/app/cli/src/commands/export_command.rs:65]: Should we support single-file export option too?

File: src/domain/core/src/services/export_service.rs
Comment (sergiimk) [inline src/domain/core/src/services/export_service.rs:21]: Would it be better to decouple ExportService and QueryService by simply accepting a `DataFrame` here instead of SQL?

File: src/infra/core/src/query_service_impl.rs
Comment (sergiimk) [inline src/infra/core/src/query_service_impl.rs:86]: Is there no way to override this in `write_parquet`? 

I would like to avoid tying `QueryService` interface so closely to DataFusion internals by surfacing such in-depth config parameters.

Comment (demusdev) [reply src/infra/core/src/query_service_impl.rs:86]: I didn't find options.
This is where it's being used: https://github.com/apache/datafusion/blob/2464703c84c400a09cc59277018813f0e797bb4e/datafusion/core/src/datasource/file_format/write/demux.rs#L150-L150
And TaskContext (SessionContext.state) comes from `write_parquet`, where new dataframe is being created: https://github.com/apache/datafusion/blob/2464703c84c400a09cc59277018813f0e797bb4e/datafusion/core/src/dataframe/parquet.rs#L85-L85
And latter comes from context we create.

If you want to have these services completely independent, we can just have session creation code (mostly)dup in new service, or move `create_session` and `create_session_with_config_overrides` to a third service and have it as a dependency for QueryService and ExportService.

File: src/domain/core/src/services/export_service.rs
Comment (demusdev) [reply src/domain/core/src/services/export_service.rs:21]: we need to create specific session context with single target partition and specific rows count per target file.

File: src/app/cli/src/commands/export_command.rs
Comment (demusdev) [reply src/app/cli/src/commands/export_command.rs:65]: yes. it's very early draft PR which I created just for visibility. I plan to add description and showcase tests

File: src/app/cli/src/cli.rs
Comment (demusdev) [reply src/app/cli/src/cli.rs:390]: I plan to add params documentation later in this PR

File: src/domain/core/src/services/export_service.rs
Comment (demusdev) [reply src/domain/core/src/services/export_service.rs]: Makes sense! I will change it.

File: src/infra/core/src/query_service_impl.rs
Comment (demusdev) [reply src/infra/core/src/query_service_impl.rs:86]: My prev statement is not quite true. I figured out how to reassemble DataFrame with hacked session config.

File: src/app/cli/src/cli.rs
Comment (zaychenko-sergei) [reply src/app/cli/src/cli.rs:390]: It's not about the docs, it's about the name

Comment (demusdev) [reply src/app/cli/src/cli.rs:390]: but name was take from ticket description: https://github.com/kamu-data/kamu-cli/issues/353
should we change it to something like `--rows-per-file`?

Comment (zaychenko-sergei) [reply src/app/cli/src/cli.rs:403]: Yes, the switch is unclear similarly. Maybe like this or maybe `--records-per-file`.

File: src/app/cli/src/commands/export_command.rs
Comment (zaychenko-sergei) [inline src/app/cli/src/commands/export_command.rs:25]: Don't we have enums like the following in other CLI commands:

```
#[derive(Debug, Clone, Copy, clap::ValueEnum)]
pub enum OutputFormat {
    /// Comma-separated values
    #[clap(name = "csv")]
    Csv,

    /// Array of Structures format
    #[clap(name = "json")]
    Json,

    /// One Json object per line - easily splittable format
    #[clap(name = "ndjson")]
    NdJson,

    /// Structure of arrays - more compact and efficient format for encoding
    /// entire dataframe
    #[clap(name = "json-soa")]
    JsonSoA,

    /// Array of arrays - compact and efficient and preserves column order
    #[clap(name = "json-aoa")]
    JsonAoA,

    /// A pretty human-readable table
    #[clap(name = "table")]
    Table,
}
```

Comment (zaychenko-sergei) [reply src/app/cli/src/commands/export_command.rs:25]: I don't mean using this particular one, but perhaps your own `ExportFormat` listing valid options.

File: src/app/cli/src/commands/export_command.rs
Comment (zaychenko-sergei) [inline src/app/cli/src/commands/export_command.rs:77]: Maybe something like this:

```   
   let path = env::current_dir()?;
   path.join(self.dataset_ref.to_string())
```   
also, should we add some file extension depending on the format?

File: src/app/cli/src/commands/sql_shell_command.rs
Comment (zaychenko-sergei) [inline src/app/cli/src/commands/sql_shell_command.rs:53]: I guess you can declare it as `const` in the module. Does not feel like it's a state.

File: src/app/cli/src/output/output_config.rs
Comment (zaychenko-sergei) [inline src/app/cli/src/output/output_config.rs:120]: And why custom display was necessary? Can't we live with the automatic version?

File: src/domain/core/src/services/errors.rs
Comment (zaychenko-sergei) [inline src/domain/core/src/services/errors.rs:22]: It feels like a boundary between domains.. 
Do we get any benefit from depending on a particular error type here?

See how it's done for `InternalError`:

```
pub type BoxedError = Box<dyn std::error::Error + Send + Sync>;

#[derive(Error, Debug)]
#[error("Internal error")]
pub struct InternalError {
    #[source]
    source: BoxedError,
    context: Option<String>,
    #[backtrace]
    backtrace: Option<Backtrace>,
}
```

File: src/domain/core/src/services/export_service.rs
Comment (zaychenko-sergei) [inline src/domain/core/src/services/export_service.rs:51]: There is `strum` crate we use in many places, maybe it can help to avoid this boilerplate?

File: src/domain/core/src/services/export_service.rs
Comment (zaychenko-sergei) [reply src/domain/core/src/services/export_service.rs:61]: There is `unimplemented!` macro that is a specific kind of panic for this situation.

File: src/app/cli/src/commands/export_command.rs
Comment (zaychenko-sergei) [reply src/app/cli/src/commands/export_command.rs:77]: As I understand, the extensions are automatically attached by the `DataFusion` writers. Did you test what happens if I specify particular extensions in the path? Wrong extensions for the selected format?

File: src/app/cli/src/output/output_config.rs
Comment (zaychenko-sergei) [reply src/app/cli/src/output/output_config.rs:120]: `strum`?

File: src/domain/core/src/services/errors.rs
Comment (zaychenko-sergei) [inline src/domain/core/src/services/errors.rs:24]: Why not to keep it where it was? It's accessible in the module namespace.

File: src/domain/core/src/services/export_service.rs
Comment (zaychenko-sergei) [inline src/domain/core/src/services/export_service.rs:28]: What confuses a bit is that is not simpliy a path, rather a prefix of the path.
Maybe some comments?

Comment (zaychenko-sergei) [inline src/domain/core/src/services/export_service.rs:28]: If we compare it to just extending `QueryService` with the `export_to_fs` method, is there any benefit in separate interface? It seems they travel together.

File: src/app/cli/src/commands/export_command.rs
Comment (demusdev) [reply src/app/cli/src/commands/export_command.rs:77]: it accepts any extension you specify. you simply need to have any. otherwise it will assume path as a dir

Comment (demusdev) [reply src/app/cli/src/commands/export_command.rs:77]: here's related datafusion writer https://github.com/apache/datafusion/blob/1fc776981496c0f0648ece8069992c65a7021c83/datafusion/core/src/datasource/file_format/write/demux.rs#L67

File: src/domain/core/src/services/export_service.rs
Comment (demusdev) [reply src/domain/core/src/services/export_service.rs:28]: it's a path. either path to file or to directory. if it's a file path (something with extension and without trailing path delimiter), DataFusion will store data into it. Otherwise, it treats path as a dir and will create files (with generated names, and extension related to export format) inside.

File: src/domain/core/src/services/export_service.rs
Comment (zaychenko-sergei) [reply src/domain/core/src/services/export_service.rs]: @sergiimk Can you explain how `DataFrame` object will help. Isn't this `DataFusion`'s specific trait?
I understand Spark and Flink have a similar concept, but I guess it's not the same interface?
Or do we plan to wrap muiltiple types with our own abstraction?

File: src/app/cli/src/output/output_config.rs
Comment (demusdev) [reply src/app/cli/src/output/output_config.rs:120]: We use `fmt` when printing error message, where we need to display supported formats. And we need custom formatter, to support all custom mappings like "NdJson" -> "ndjson", "JsonSoA" -> "json-soa"

File: src/domain/core/src/services/export_service.rs
Comment (demusdev) [reply src/domain/core/src/services/export_service.rs:28]: I think logically it's a different things, `QueryService` is for preparing/reading data, and `ExportService` is for writing. When I started working on code, I was expecting something much more complex than just a few lines of code so created a separate service.
But with the current code state, having it separated or not is more matter of taste. I just left as it was.

**Review by zaychenko-sergei — APPROVED** [review]

**Comment by zaychenko-sergei** [comment]
@demusdev wait for @sergiimk 's approval.

File: src/domain/core/src/services/export_service.rs
Comment (sergiimk) [reply src/domain/core/src/services/export_service.rs]: `DataFrame` can wrap Arrow buffers and our long-term goal is to make Arrow the only data interchange format in Kamu. 

There is a high chance Spark and other engines will soon natively support Arrow format. If they don't we'll imlement adapters. But the idea is that `DataFrame` will be able to represent output data from any engine in ODF, not only DataFusion.

File: src/domain/core/src/services/export_service.rs
Comment (sergiimk) [inline src/domain/core/src/services/export_service.rs:29]: Let's add `ExportOpts` struct and move this parameter there. This way users will be able to pass `ExportOpts::default()` where they don't care about the details, and we will be able to add more options without breaking every place where this method is called.

This is a pattern a lot of our services follow.

File: src/domain/core/src/services/export_service.rs
Comment (sergiimk) [inline src/domain/core/src/services/export_service.rs:68]: Aren't all DF errors still internal errors? Why do we need to differentiate them?

**Review by sergiimk — APPROVED** [review]
Left a few small suggestions but looks good overall :ship:

File: src/app/cli/src/cli.rs
Comment (demusdev) [reply src/app/cli/src/cli.rs:400]: I think it makes sense to let the engine to choose default, so I voting for the first option

File: resources/cli-reference.md
Comment (demusdev) [reply resources/cli-reference.md]: done

File: src/domain/core/src/services/export_service.rs
Comment (demusdev) [reply src/domain/core/src/services/export_service.rs:83]: `ExportError::Internal` is thrown when we failed to calculate exported rows.
For all other cases we throw `DataFusionError`. I took QueryService as an example, so new command and changes to `sql` one conforms existent interface. 
With this approach we sometimes may show user not useful details. But if we want to improve this, we need to be careful with filtering DataFusion errors in order to provide users with details when their sql is invalid, or they cannot store data to a specific location. I think we need to make changes to both Query and External service at the same time. And if we'll go with this direction, I guess it makes sense to implement it in scope of separate ticket.

Comment (demusdev) [reply src/domain/core/src/services/export_service.rs:83]: For the context, here're some of DataFusion errors exposed to users.
Triggered from QueryService:
```bash
❯ kamu sql --command "select random_field_name from 'test.data2'" --output-format parquet --output-path ./out/parquet/1
Error:
  0: DataFusion error
  1: Schema error
  2: No field named random_field_name. Valid fields are "test.data2".offset, "test.data2".op, "test.data2".system_time, "test.data2".event_time, "test.data2".uuid1.
  
❯ kamu sql --command "select * from 'test.data2' qwe 212" --output-format parquet --output-path /nix/1.parquet
Error:
  0: DataFusion error
  1: SQL error: ParserError("Expected end of statement, found: 212")
  2: sql parser error: Expected end of statement, found: 212

❯ kamu sql --command "select * from 'does.not.exists'" --output-format parquet --output-path ./out/parquet/1
Error:
  0: DataFusion error
  1: Error during planning: table 'kamu.kamu.does.not.exists' not found
```
Triggered from ExportService:
```bash❯ kamu sql --command "select 1/0" --output-format parquet --output-path ./out/parquet/1
Error:
  0: DataFusion error
  1: Arrow error
  2: Divide by zero error
  
❯ kamu sql --command "select * from 'test.data2'" --output-format parquet --output-path /out/parquet/1
Error:
  0: DataFusion error
  1: IO error
  2: Generic LocalFileSystem error
  3: Unable to create dir /out/parquet/1
  4: Read-only file system (os error 30)
  
❯ kamu sql --command "select * from 'test.data2'" --output-format parquet --output-path /nix/1.parquet
Error:
  0: DataFusion error
  1: IO error
  2: Generic LocalFileSystem error
  3: Unable to open file /nix/1.parquet#1
  4: Permission denied (os error 13)  
  
❯ kamu sql --command "select * from 'test.data2' limit -1" --output-format parquet --output-path /nix/1.parquet
Error:
  0: DataFusion error
  1: Optimizer rule 'eliminate_limit' failed
caused by
Error during planning: LIMIT must be >= 0, '-1' was provided
  2: Error during planning: LIMIT must be >= 0, '-1' was provided
```

Comment (sergiimk) [reply src/domain/core/src/services/export_service.rs:83]: All errors you provided as examples above look like planning-phase errors to me, so If I'm not mistaken they should all be raised from the `QueryServiceImpl` while we are setting up the logical plan, not from `ExportServiceImpl`.

Here it makes sense for `QueryService` to return `DataFusionError` instead of `InternalError` because very likely all errors will be caused by the invalid user input.

This logical plan in form of a `DataFrame` is then passed into the `ExportService` where we add a few extra steps:
- repartition + file sync - a planning error on at phase will indicate we badly messed up the code
- finally we execute the plan - an error at this phase will likely be IO-related (out of space, permissions etc).

In both cases there isn't much what user can do about them. And the latter execution-phase errors when used in `kamu-node` context can reveal details of internal infrastructure (e.g. some S3 bucket names or object keys).

Comment (demusdev) [reply src/domain/core/src/services/export_service.rs:83]: Some errors like division by zero and invalid limit value are spotted on execution phase (in ExportService). I believe there are a few more useful ones I didn't spot. And also IO errors in example above looks useful as well. I'd preserve them either as DataFusion errors, or create specific ones, and would map corresponding DataFusion errors to that new ones.

**Pull request merged**
